engine/table/src/main/java/io/deephaven/engine/table/impl/indexer/DataIndexer.java

```diff
@@ -0,0 +1,256 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.engine.table.impl.indexer;
+
+import io.deephaven.engine.rowset.RowSet;
+import io.deephaven.engine.rowset.TrackingRowSet;
+import io.deephaven.engine.table.ColumnSource;
+import io.deephaven.engine.table.DataIndex;
+import io.deephaven.engine.table.Table;
+import io.deephaven.engine.table.impl.QueryTable;
+import io.deephaven.engine.table.impl.dataindex.TableBackedDataIndexImpl;
+import org.apache.commons.lang3.mutable.MutableBoolean;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Indexer that provides single and multi-column clustering indexes for a table, linked to a {@link TrackingRowSet}.
+ * 
+ * @apiNote DataIndexers should not be used after the host {@link TrackingRowSet} has been {@link RowSet#close()
+ *          closed}.
+ */
+public class DataIndexer implements TrackingRowSet.Indexer {
+    public static DataIndexer of(TrackingRowSet rowSet) {
+        return rowSet.indexer(DataIndexer::new);
+    }
+
+    private final TrackingRowSet rowSet;
+
+    /**
+     * The root of a pseudo-trie of index caches. This is a complicated structure but has the strong benefit of allowing
+     * GC of data indexes belonging to GC'd column sources.
+     */
+    WeakHashMap<ColumnSource<?>, DataIndexCache> dataIndexes;
+
+    private static class DataIndexCache {
+
+        /** The index at this level. */
+        @Nullable
+        private volatile DataIndex index;
+
+        /** The sub-indexes below this level. */
+        private final WeakHashMap<ColumnSource<?>, DataIndexCache> dataIndexes;
+
+        DataIndexCache(@Nullable final DataIndex index) {
+            this.index = index;
+            this.dataIndexes = new WeakHashMap<>();
+        }
+    }
+
+    private DataIndexer(@NotNull final TrackingRowSet rowSet) {
+        this.rowSet = rowSet;
+        this.dataIndexes = new WeakHashMap<>();
+    }
+
+    public boolean hasDataIndex(final Table table, final String... keyColumnNames) {
+        final DataIndex index = getDataIndex(table, keyColumnNames);
```

Comment:
It surprises me that this doesn't remap to column sources and call `hasDataIndex`.